api/src/components/coauthor/controller.ts

```diff
@@ -0,0 +1,204 @@
+import * as response from 'lib/response';
+import * as publicationService from 'publication/service';
+import * as coAuthorService from './service';
+import * as I from 'interface';
+
+export const create = async (
+    event: I.AuthenticatedAPIRequest<I.CreateCoAuthorRequestBody, undefined, I.CreateCoAuthorPathParams>
+): Promise<I.JSONResponse> => {
+    try {
+        const publication = await publicationService.get(event.pathParameters.id);
+
+        // Does the publication exist?
+        if (!publication) {
+            return response.json(404, {
+                message: 'This publication does not exist.'
+            });
+        }
+
+        // Is this user the author of the publication?
+        if (publication?.user.id !== event?.user.id) {
+            return response.json(403, {
+                message: 'You do not have the right permissions for this action.'
+            });
+        }
+
+        // Is the publication in draft?
+        if (publication.currentStatus === 'LIVE') {
+            return response.json(403, {
+                message: 'This publication is LIVE and therefore cannot be edited.'
+            });
+        }
+
+        const isUserAlreadyCoAuthor = await coAuthorService.isUserAlreadyCoAuthor(
+            event.body.email,
+            event.pathParameters.id
+        );
+
+        if (isUserAlreadyCoAuthor) {
+            return response.json(409, { message: 'This email has already been added as a co-author.' });
+        }
+
+        const coAuthor = await coAuthorService.create(event.body.email, event.pathParameters.id);
+
+        return response.json(201, coAuthor);
+    } catch (err) {
+        console.log(err);
+        return response.json(500, { message: 'Unknown server error.' });
+    }
+};
+
+export const remove = async (
+    event: I.AuthenticatedAPIRequest<undefined, undefined, I.DeleteCoAuthorPathParams>
+): Promise<I.JSONResponse> => {
+    try {
+        const publication = await publicationService.get(event.pathParameters.id);
+
+        // Does the publication exist?
+        if (!publication) {
+            return response.json(404, {
+                message: 'This publication does not exist.'
+            });
+        }
+
+        // Is this user the author of the publication?
+        if (publication.user.id !== event.user.id) {
+            return response.json(403, {
+                message: 'You do not have the right permissions for this action.'
+            });
+        }
+
+        // Is the publication in draft?
+        if (publication.currentStatus === 'LIVE') {
```

Comment:
Same as the above, as this logic is duped, maybe a middleware or similar can be setup to run these two checks

Reply:
Same as above